src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+    b[6] = ((r[4] << 4) | (r[5] >> 2)) as u8;
+    b[7] = ((r[5] << 6) | r[6]) as u8;
+
+    b[8] = ((r[7] << 2) | (r[8] >> 4)) as u8;
+    b[9] = ((r[8] << 4) | (r[9] >> 2)) as u8;
+    b[10] = ((r[9] << 6) | r[10]) as u8;
+
+    b[11] = ((r[11] << 2) | (r[12] >> 4)) as u8;
+    b[12] = ((r[12] << 4) | (r[13] >> 2)) as u8;
+    b[13] = ((r[13] << 6) | r[14]) as u8;
+
+    return 14;
+}
+
+fn b_u32(b: &[u8], i: usize) -> u32 {
+    b[i] as u32
+}
+
+// 0011 1111
+const SIX_BITS: u32 = 0x3f; // 0x3fu8
+
+// Unpack a 14-byte block into 4 by 4 16-bit pixels.
+fn unpack14(b: &[u8], s: &mut [u16; 16]) {
+    // TODO: b slice should be &mut [u8; 14], but rust doesn't support sized slice.
+    assert_eq!(b.len(), 14);
+    assert_ne!(b[2], 0xfc);
+
+    s[0] = ((b_u32(&b, 0) << 8) | b_u32(&b, 1)) as u16;
+
+    let shift = (b_u32(&b, 2) >> 2);
+    let bias = 0x20 << shift;
+
+    s[4] = (s[0] as u32 + ((((b_u32(&b, 2) << 4) | (b_u32(&b, 3) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[8] = (s[4] as u32 + ((((b_u32(&b, 3) << 2) | (b_u32(&b, 4) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[12] = (s[8] as u32 + ((b_u32(&b, 4) & SIX_BITS) << shift) - bias) as u16;
+
+    s[1] = (s[0] as u32 + ((b_u32(&b, 5) >> 2) << shift) - bias) as u16;
+    s[5] = (s[4] as u32 + ((((b_u32(&b, 5) << 4) | (b_u32(&b, 6) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[9] = (s[8] as u32 + ((((b_u32(&b, 6) << 2) | (b_u32(&b, 7) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[13] = (s[12] as u32 + ((b_u32(&b, 7) & SIX_BITS) << shift) - bias) as u16;
+
+    s[2] = (s[1] as u32 + ((b_u32(&b, 8) >> 2) << shift) - bias) as u16;
+    s[6] = (s[5] as u32 + ((((b_u32(&b, 8) << 4) | (b_u32(&b, 9) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[10] = (s[9] as u32 + ((((b_u32(&b, 9) << 2) | (b_u32(&b, 10) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[14] = (s[13] as u32 + ((b_u32(&b, 10) & SIX_BITS) << shift) - bias) as u16;
+
+    s[3] = (s[2] as u32 + ((b_u32(&b, 11) >> 2) << shift) - bias) as u16;
+    s[7] = (s[6] as u32 + ((((b_u32(&b, 11) << 4) | (b_u32(&b, 12) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[11] = (s[10] as u32 + ((((b_u32(&b, 12) << 2) | (b_u32(&b, 13) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[15] = (s[14] as u32 + ((b_u32(&b, 13) & SIX_BITS) << shift) - bias) as u16;
+
+    for i in 0..16 {
+        if (s[i] & 0x8000) != 0 {
+            s[i] &= 0x7fff;
+        } else {
+            s[i] = !s[i];
+        }
+    }
+}
+
+// Unpack a 3-byte block into 4 by 4 identical 16-bit pixels.
+fn unpack3(b: &[u8], s: &mut [u16; 16]) {
+    // TODO: b slice should be &mut [u8; 3], but rust doesn't support sized slice.
+    assert_eq!(b[2], 0xfc);
+
+    s[0] = (((b[0] as u32) << 8) | (b[1] as u32)) as u16;
+
+    if (s[0] & 0x8000) != 0 {
+        s[0] &= 0x7fff;
+    } else {
+        s[0] = !s[0];
+    }
+
+    for i in 0..16 {
+        s[i] = s[0];
+    }
```

Comment:
Any idea on how to change this to a real rusty approach?

Reply:
Sadly, there is no convenience function for this in Rust `std`. 

Here is how I would do it:
```rust
let (first, rest) = s.split_first_mut().unwrap();  // split off first item, gain mutable access to both (only fails if slice would be empty)
for element in rest { *element = *first; } // feels like it should be a predefined function, but will always be optimized, no worries
```

Or maybe:
```rust
let first = s[0]; 
for element in &mut s[1..] { *element = first; }
```

1. it skips the first element when writing (which is not a big performance win, but delivers intent). 
2. In this PR in general, many index-based loops could be replaced with direct reference loops. This is easier to read and can be optimized by the compiler. So, I would always prefer `for x in &slice { }` over `for i in range { }`. You can use `for (elem, index) in slice.iter().enumerate() { }` if you also need the index for computing the element. This is also how the general Rust community prefers the loops :)

Reply:
Now that I look at the context, the first version does not make sense, because we never needed to store the temporary value in an array.

```rust
    assert_eq!(b[2], 0xfc);
    let sample = (((b[0] as u32) << 8) | (b[1] as u32)) as u16;
    let sample = if (sample & 0x8000) != 0 { sample & 0x7fff } else { !sample };
    for element in s { *element = sample; }
```

Reply:
I will take the review one by one. Actually, the whole code hint is now easier to understand than focusing on the loop only:

```rust
// Unpack a 3-byte block `b` into 4 by 4 identical 16-bit pixels in `s` array.
fn unpack3(b: &[u8], s: &mut [u16; 16]) {
    assert_eq!(b[2], 0xfc);

    // Get the pixel value (convert the 2 bytes to u16).
    let mut value = (((b[0] as u32) << 8) | (b[1] as u32)) as u16;

    if (value & 0x8000) != 0 {
        value &= 0x7fff;
    } else {
        value = !value;
    }

    s.fill(value);
}
```

Note the use of `fill()`.

I think this version is more clear: Extract the float value and put it everywhere.

Reply:
Here, `s` is an `array` (`[u16; 16]`), not a slice. `value` is an `u16`. So `array.fill()` is the proper way.

Reply:
Excellent! I wonder why this method is not available for all slices, quite disappointing to have it only for arrays